### PR TITLE
8307363: TextFlow.underlineShape()

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Adding TextFlow.underlineShape() to add support for a spellchecker-like decoration, using 
```
getRange(start, end, TextLayout.TYPE_UNDERLINE);
```

which mirrors an existing method in Text with exactly the same signature.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8307861](https://bugs.openjdk.org/browse/JDK-8307861) to be approved
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8307363](https://bugs.openjdk.org/browse/JDK-8307363): TextFlow.underlineShape()
 * [JDK-8307861](https://bugs.openjdk.org/browse/JDK-8307861): TextFlow.underlineShape() (**CSR**)

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1127/head:pull/1127` \
`$ git checkout pull/1127`

Update a local copy of the PR: \
`$ git checkout pull/1127` \
`$ git pull https://git.openjdk.org/jfx.git pull/1127/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1127`

View PR using the GUI difftool: \
`$ git pr show -t 1127`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1127.diff">https://git.openjdk.org/jfx/pull/1127.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1127#issuecomment-1535489091)

### CODE REVIEW

**Comment by andy-goryachev-oracle** [comment]
@nlisker would you be able to take a look at this please?

**Comment by nlisker** [comment]
I'm not familiar enough with these classes to review. Is there something specific you want me to look at?

**Comment by kevinrushforth** [comment]
Maybe @prrace can be the second reviewer?

**Review by kevinrushforth — COMMENTED** [review]
The API looks good. Go ahead and create the CSR. The implementation looks correct, as does the manual test addition to MonkeyTester. Can you also add an automated test?

**Comment by kevinrushforth** [comment]
It's definitely affected by GC, so I'd start by looking there. I can change whether the the test passes or fails by changing the heap size or adding a call to gc before each test.

**Comment by andy-goryachev-oracle** [comment]
> UPDATE: THIS COMMENT WAS MEANT FOR ANOTHER PR.

we've all been there!  :-)

**Comment by andy-goryachev-oracle** [comment]
CSR created https://bugs.openjdk.org/browse/JDK-8307861

File: tests/system/src/test/java/test/javafx/scene/text/TextFlowNodeTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        PlatformImpl.startup(() -> {
+            startupLatch.countDown();
+        });
+        try {
+            if (!startupLatch.await(5, TimeUnit.SECONDS)) {
+                Assert.fail("Timeout waiting for FX runtime to start");
+            }
+        } catch (InterruptedException ex) {
+            Assert.fail("Unexpected exception: " + ex);
+        }
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/scene/text/TextFlowNodeTest.java:57]: Can this be replaced with `Util.startup()`?

File: tests/system/src/test/java/test/javafx/scene/text/TextFlowNodeTest.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    public void testUnderlineShape() {
+        Text t1 = new Text("one ");
+        t1.setFont(new Font("Monospaced Regular", 16));
+        Text t2 = new Text("two.");
+        TextFlow f = new TextFlow(t1, t2);
+
+        // underline 0,0 must be empty
+        PathElement[] p = f.underlineShape(0, 0);
+        Assert.assertNotNull(p);
+        Assert.assertEquals(p.length, 0);
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/scene/text/TextFlowNodeTest.java:70]: Arguments are switched (expected value comes first).

**Review by kevinrushforth — COMMENTED** [review]
Test looks good with a couple comments. I'll give it a quick test drive.

File: tests/system/src/test/java/test/javafx/scene/text/TextFlowNodeTest.java (hunk 1 above)
Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/javafx/scene/text/TextFlowNodeTest.java:57]: "shoemaker's children go barefoot"

**Review by kevinrushforth — APPROVED** [review]
Looks good.

**Comment by kevinrushforth** [comment]
@prrace Can you re-review this when you get a chance?

**Comment by kevinrushforth** [comment]
@prrace Can you re-review this when you get a chance?

**Pull request closed without merging**